3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt

```diff
@@ -0,0 +1,325 @@
+        }
+    }
+
+    @Throws(ComponentException::class)
+    fun createEncodedFingerprint(authenticationRequestParameters: AuthenticationRequestParameters): String {
+        val fingerprintJson = JSONObject()
+        try {
+            // TODO getMessageVersion is not used?
+            fingerprintJson.put("sdkAppID", authenticationRequestParameters.sdkAppID)
+            fingerprintJson.put("sdkEncData", authenticationRequestParameters.deviceData)
+            fingerprintJson.put("sdkEphemPubKey", JSONObject(authenticationRequestParameters.sdkEphemeralPublicKey))
+            fingerprintJson.put("sdkReferenceNumber", authenticationRequestParameters.sdkReferenceNumber)
+            fingerprintJson.put("sdkTransID", authenticationRequestParameters.sdkTransactionID)
+        } catch (e: JSONException) {
+            throw ComponentException("Failed to create encoded fingerprint", e)
+        }
+        return Base64Encoder.encode(fingerprintJson.toString())
+    }
+
+    private fun createChallengeParameters(challenge: ChallengeToken): ChallengeParameters {
+        // TODO referenceNumber, URL and version are not used?
+        val challengeParameters = ChallengeParameters()
+        challengeParameters.set3DSServerTransactionID(challenge.threeDSServerTransID)
+        challengeParameters.acsTransactionID = challenge.acsTransID
+        challengeParameters.acsRefNumber = challenge.acsReferenceNumber
+        challengeParameters.acsSignedContent = challenge.acsSignedContent
+        return challengeParameters
+    }
+
+    private fun closeTransaction(context: Context) {
+        mTransaction?.close()
+        mTransaction = null
+        try {
+            ThreeDS2Service.INSTANCE.cleanup(context)
+        } catch (e: SDKNotInitializedException) {
+            // no problem
+        }
+    }
+
+    @Throws(ComponentException::class)
+    fun createFingerprintDetails(encodedFingerprint: String?): JSONObject {
+        val fingerprintDetails = JSONObject()
+        try {
+            fingerprintDetails.put(FINGERPRINT_DETAILS_KEY, encodedFingerprint)
+        } catch (e: JSONException) {
+            throw ComponentException("Failed to create fingerprint details", e)
+        }
+        return fingerprintDetails
+    }
+
+    @Throws(ComponentException::class)
+    private fun createChallengeDetails(completionEvent: CompletionEvent): JSONObject {
+        val challengeDetails = JSONObject()
+        try {
+            val challengeResult = ChallengeResult.from(completionEvent)
+            challengeDetails.put(CHALLENGE_DETAILS_KEY, challengeResult.payload)
+        } catch (e: JSONException) {
+            throw ComponentException("Failed to create challenge details", e)
+        }
+        return challengeDetails
+    }
+
+    companion object {
+        val TAG = LogUtil.getTag()
+
+        val PROVIDER: ActionComponentProvider<Adyen3DS2Component> = ActionComponentProviderImpl(
```

Comment:
Does this need `@JvmField`? Same question for the other companion variables?

Reply:
Good catch